fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingMutuallyByRecords.java

```diff
@@ -0,0 +1,529 @@
+                                          lastResult.get().get().getPrimaryKey() :
+                                          rangeEnd)
+                    .thenCompose(cont -> insertRanges(store.ensureContextActive(), targetRangeSets, packOrNull(rangeStart), packOrNull(cont))
+                            .thenApply(ignore -> !allRangesExhausted(cont, rangeEnd)));
+        });
+    }
+
+    @Nullable
+    private static Range fullyUnBuiltRange(List<Range> missingRanges, Range fragmentRange) {
+        // Return the fragmentRange 'as is' if it appears to be fully un-built in the
+        // missing ranges list. Else return null.
+        // * Assuming that the missing ranges are sorted and squashed
+        // * Optimize for a heavily fragmented missing ranges. (Should we perform a binary search?)
+        for (Range range: missingRanges) {
+            if (ByteArrayUtil.compareUnsigned(range.end, fragmentRange.end) >= 0) {
+                // missing range is beyond fragment
+                if (ByteArrayUtil.compareUnsigned(range.begin, fragmentRange.begin) <= 0) {
+                    return notEmptyRange(fragmentRange);
+                }
+                break;
+            }
+        }
+        return null;
+    }
+
+    @Nullable
+    private static Range partlyUnBuiltRange(List<Range> missingRanges, Range fragmentRange) {
+        // Return null if the range appears to be fully built in the missing ranges list. Else
+        // return the un-built range of (fragmentRange ∩ missingRanges)
+        // * Assuming that the missing ranges are sorted and squashed
+        // * Less optimized than the fullyUnBuiltRanges (this function is expected to be called only a handful
+        //   of times at the end of the indexing process)
+        for (Range range: missingRanges) {
+            if (ByteArrayUtil.compareUnsigned(range.begin, fragmentRange.end) > 0 ||
+                    ByteArrayUtil.compareUnsigned(range.end, fragmentRange.begin) < 0) {
+                // no range overlap
+                continue;
+            }
+            // Here: some range overlap, shrink to fragmentRange ∩ range
+            byte[] begin = ByteArrayUtil.compareUnsigned(range.begin, fragmentRange.begin) >= 0 ?
+                           range.begin : fragmentRange.begin;
+            byte[] end = ByteArrayUtil.compareUnsigned(range.end, fragmentRange.end) <= 0 ?
+                         range.end : fragmentRange.end;
+            if (ByteArrayUtil.compareUnsigned(begin, end) < 0) {
+                // if non-empty range
+                return notEmptyRange(new Range(begin, end));
+            }
+        }
+        return null;
+    }
+
+    @Nullable
+    private static Range notEmptyRange(Range range) {
+        // check the (rare, maybe test only) case of empty ranges
+        return ByteArrayUtil.compareUnsigned(range.begin, range.end) >= 0 ?
+               null : range;
+    }
+
+    List<Range> sortAndSquash(List<Range> ranges) {
```

Comment:
This method, `fullyUnBuiltRange` and `partlyUnBuiltRange` seem like they might be complicated enough to warrant some unit tests.

Reply:
✅